src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp

```diff
@@ -189,44 +192,106 @@ namespace OpenLoco::World::MapGenerator
+    {
+        for (auto& pos : World::getDrawableTileRange())
+        {
+            const bool tileIsMarked = heightMap.isMarkerSet({ pos.x, pos.y });
+            if (requireMark != tileIsMarked)
+                continue;
+
+            auto tile = TileManager::get(pos);
+            auto* surface = tile.surface();
+            if (surface == nullptr)
+                continue;
+
+            surface->setTerrain(surfaceStyle);
+            auto res = getRandomTerrainVariation(*surface);
+            if (res)
+                surface->setVariation(*res);
+        }
+    }
+
     // 0x0046A379
     static void generateTerrainFarFromWater(HeightMap& heightMap, uint8_t surfaceStyle)
     {
-        _heightMap = heightMap.data();
-        registers regs;
-        regs.ebx = surfaceStyle;
-        call(0x0046A379, regs);
-        _heightMap = nullptr;
+        heightMap.resetMarkerFlags();
+
+        // Mark tiles near water
+        auto seaLevel = getGameState().seaLevel;
+        for (auto pos : getWorldRange())
+        {
+            auto height = heightMap.getHeight({ pos.x, pos.y });
+            if (height > seaLevel)
+                continue;
+
+            for (auto lookaheadPos : getClampedRange(pos, pos + TilePos2(50, 50)))
+                heightMap.setMarker({ lookaheadPos.x, lookaheadPos.y });
+        }
+
+        // Apply surface style to tiles that have *not* been marked
+        applySurfaceStyleToMarkedTiles(heightMap, surfaceStyle, false);
     }
 
     // 0x0046A439
     static void generateTerrainNearWater(HeightMap& heightMap, uint8_t surfaceStyle)
     {
-        _heightMap = heightMap.data();
-        registers regs;
-        regs.ebx = surfaceStyle;
-        call(0x0046A439, regs);
-        _heightMap = nullptr;
+        heightMap.resetMarkerFlags();
+
+        // Mark tiles near water
+        auto seaLevel = getGameState().seaLevel;
+        for (auto pos : getWorldRange())
+        {
+            auto height = heightMap.getHeight({ pos.x, pos.y });
+            if (height < seaLevel)
+                continue;
+
+            for (auto lookaheadPos : getClampedRange(pos, pos + TilePos2(50, 50)))
+                heightMap.setMarker({ lookaheadPos.x, lookaheadPos.y });
+        }
+
+        // Apply surface style to tiles that have been marked
+        applySurfaceStyleToMarkedTiles(heightMap, surfaceStyle, true);
     }
 
     // 0x0046A5B3
     static void generateTerrainOnMountains(HeightMap& heightMap, uint8_t surfaceStyle)
     {
-        _heightMap = heightMap.data();
-        registers regs;
-        regs.ebx = surfaceStyle;
-        call(0x0046A5B3, regs);
-        _heightMap = nullptr;
+        heightMap.resetMarkerFlags();
+
+        // Mark tiles above mountain level
+        for (auto pos : getWorldRange())
+        {
+            auto height = heightMap.getHeight({ pos.x, pos.y });
+            if (height <= kMountainTerrainHeight)
```

Comment:
I really think we need to have a comment here confirming this is `<=` and below is `<` so a future dev doesn't look at it and go "oh those are different, those should be the same"